tahoe_idp/models.py

```diff
@@ -0,0 +1,147 @@
+            'user': user,
+            'magiclink': self.generate_url(request),
+            'expiry': self.expiry,
+            'ip_address': self.ip_address,
+            'created': self.created,
+            'require_same_ip': magiclink_settings.REQUIRE_SAME_IP,
+            'require_same_browser': magiclink_settings.REQUIRE_SAME_BROWSER,
+            'token_uses': magiclink_settings.TOKEN_USES,
+            'style': magiclink_settings.EMAIL_STYLES,
+        }
+        plain = render_to_string(magiclink_settings.EMAIL_TEMPLATE_NAME_TEXT, context)
+        html = render_to_string(magiclink_settings.EMAIL_TEMPLATE_NAME_HTML, context)
+        send_mail(
+            subject=magiclink_settings.EMAIL_SUBJECT,
+            message=plain,
+            recipient_list=[user.email],
+            from_email=djsettings.DEFAULT_FROM_EMAIL,
+            html_message=html,
+        )
+
+    def validate(
+        self,
+        request: HttpRequest,
+        email: str = '',
+    ) -> AbstractUser:
+        if magiclink_settings.EMAIL_IGNORE_CASE and email:
+            email = email.lower()
+
+        if magiclink_settings.VERIFY_INCLUDE_EMAIL and self.email != email:
+            raise MagicLinkError('Email address does not match')
+
+        if timezone.now() > self.expiry:
+            self.disable()
+            raise MagicLinkError('Magic link has expired')
+
+        if magiclink_settings.REQUIRE_SAME_IP:
+            client_ip = get_client_ip(request)
+            if client_ip and magiclink_settings.ANONYMIZE_IP:
+                client_ip = client_ip[:client_ip.rfind('.')+1] + '0'
+            if self.ip_address != client_ip:
+                self.disable()
+                raise MagicLinkError('IP address is different from the IP '
+                                     'address used to request the magic link')
+
+        if magiclink_settings.REQUIRE_SAME_BROWSER:
+            cookie_name = 'magiclink{pk}'.format(pk=self.pk)
+            if self.cookie_value != request.COOKIES.get(cookie_name):
+                self.disable()
+                raise MagicLinkError('Browser is different from the browser '
+                                     'used to request the magic link')
+
+        if self.times_used >= magiclink_settings.TOKEN_USES:
+            self.disable()
+            raise MagicLinkError('Magic link has been used too many times')
+
+        user = User.objects.get(email=self.email)
+
+        if not magiclink_settings.ALLOW_SUPERUSER_LOGIN and user.is_superuser:
+            self.disable()
+            raise MagicLinkError(
+                'You can not login to a super user account using a magic link')
+
+        if not magiclink_settings.ALLOW_STAFF_LOGIN and user.is_staff:
+            self.disable()
+            raise MagicLinkError(
+                'You can not login to a staff account using a magic link')
+
+        return user
+
+
+class MagicLinkUnsubscribe(models.Model):
+    email = models.EmailField()
```

Comment:
maybe remove this, we don't need the `create` workflow, hence the `unsubscribe` isn't needed